Add mappings for reactive disclosure elements

duncandewhurst opened the pull request:
Closes #220 

Depends on #294

A couple of things to note:

* I've documented mappings from OCDS, even where document type codes don't exist in that standard, in case they are added in the future
* The structure of the reactive disclosures section of the CoST IDS differs from the structure of the proactive disclosure section. Rather than trying to align the reactive disclosures with the structure of the proactive disclosures, I've reflected the different structures in the mapping tables, as I think that will be easiest to follow for CoST implementers

duncandewhurst committed:
add reactive disclosure mappings

duncandewhurst committed:
add OCDS mappings for reactive disclosures

duncandewhurst commented:
@pindec please could you review. Thanks!

pindec commented on mapping/reactive-project-level-identification-preparation.csv:
```diff
@@ -0,0 +1,46 @@
+
+* Set its ``.period.endDate`` to the last day of the year, e.g ‘2020-12-31T23:59:59Z",*Copy the contents of* ``/planning/budget/budgetBreakdown`` *in each contracting process to* ``budget.BudgetBreakdown``.
+Project brief or Feasibility study,"Documentation of feasibility studies carried out for this contracting process or project, providing information on net benefits or costs of the proposed goods, works or services.",".. markdown::
+  [Add a project document](#add-a-project-document)and set ``.documentType`` to 'feasibilityStudy'.",*Check the* ``/planning/documents`` *array for* ``Document`` *objects with* ``.type`` *set to* ‘feasibilityStudy’ *and copy each object to the* ``documents`` *array.*
```
Remove full stop in 'feasibilityStudy'." for consistency.

duncandewhurst replied:
Fixed in https://github.com/open-contracting/infrastructure/pull/295/commits/68ca204e39c4afbbc1b154acc4d4232bb7cbfde2

pindec commented on mapping/reactive-project-level-completion.csv:
```diff
@@ -0,0 +1,24 @@
+
+  **Data**
+
+  See [implementation progress reports](#implementation-progress-reports).","*Optionally, check the* ``/contracts/implementation`` *array for* ``Document`` *objects with* ``.type`` *set to* ‘physicalProgressReport’ or ‘financialProgressReport *and copy each object to the* ``.summary.documents`` *array.*
```
Remove full stop in -reports)." for consistency.

duncandewhurst replied:
Fixed in https://github.com/open-contracting/infrastructure/pull/295/commits/68ca204e39c4afbbc1b154acc4d4232bb7cbfde2

pindec commented on mapping/reactive-project-level-completion.csv:
```diff
@@ -0,0 +1,24 @@
+
+*Optionally, copy the contents of the* ``/planning/forecasts`` *array to* ``forecasts`` *and copy the contents of the* ``/contracts/implementation/metrics`` *array to* ``metrics``."
+Budget amendment decision,Document accrediting the approval of budget expansion or extension for this project.,".. markdown::
+  [Add a project document](#add-a-project-document) and set ``.documentType`` to 'budgetAmendmentApproval``",*Check the* ``/planning/documents`` *array for* ``Document`` *objects with* ``.type`` *set to* ‘budgetAmendmentApproval’ *and copy each object to the* ``documents`` *array.*
```
'budgetAmendmentApproval``" -> 'budgetAmendmentApproval'"

duncandewhurst replied:
Fixed in https://github.com/open-contracting/infrastructure/pull/295/commits/68ca204e39c4afbbc1b154acc4d4232bb7cbfde2

pindec commented on mapping/reactive-process-level-contract.csv:
```diff
@@ -0,0 +1,20 @@
+
+  * set its ``.role`` to 'supplier'
+  * set its ``.identifier.id`` to the registration number
+  * set its ``.identifier.scheme`` to the code from http://org-id.guide/ for the register the identifier is drawn from
```
Add link to org-id.guide?

duncandewhurst replied:
Added in https://github.com/open-contracting/infrastructure/pull/295/commits/68ca204e39c4afbbc1b154acc4d4232bb7cbfde2

pindec commented on docs/cost/index.md:
```diff
@@ -33,43 +33,56 @@ You can read more about the Infrastructure Data Standard in [CoST Guidance Note
+The following tables document two mappings:
 
-* How each element of the CoST Infrastructure Data Standard can be represented as **structured data** using the [Open Contracting for Infrastructure Data Standard](../projects/index), in the 'Mapping to OC4IDS' column.
+* The **CoST IDS to OC4IDS** mapping describes how to represent each element of the CoST IDS as structured data using OC4IDS. Use this mapping mapping if you already collect data according to the CoST IDS and you want to publish your data using OC4IDS, or if you want to make sure that your OC4IDS publication conforms to the CoST IDS.
```
Add link to the mapping table  #cost-ids-to-oc4ids-mapping?

pindec replied:
Double mapping in "Use this mapping mapping if you already collect"

duncandewhurst replied:
Done in https://github.com/open-contracting/infrastructure/pull/295/commits/68ca204e39c4afbbc1b154acc4d4232bb7cbfde2

pindec commented on docs/cost/index.md:
```diff
@@ -33,43 +33,56 @@ You can read more about the Infrastructure Data Standard in [CoST Guidance Note
 
-* How each element of the CoST Infrastructure Data Standard can be represented as **structured data** using the [Open Contracting for Infrastructure Data Standard](../projects/index), in the 'Mapping to OC4IDS' column.
+* The **CoST IDS to OC4IDS** mapping describes how to represent each element of the CoST IDS as structured data using OC4IDS. Use this mapping mapping if you already collect data according to the CoST IDS and you want to publish your data using OC4IDS, or if you want to make sure that your OC4IDS publication conforms to the CoST IDS.
+* The **OCDS to OC4IDS** mapping describes how to use OCDS data to populate the sections of an OC4IDS file which relate to the CoST IDS. Use this mapping if you have access to OCDS data on infrastructure contracting processes and you want to create a summary by project in OC4IDS format, or if you want to check which CoST IDS elements your OCDS data covers.
```
Add link to mapping table #ocds-to-oc4ids-mapping?

duncandewhurst replied:
Done in https://github.com/open-contracting/infrastructure/pull/295/commits/68ca204e39c4afbbc1b154acc4d4232bb7cbfde2

pindec commented on docs/cost/index.md:
````diff
@@ -33,43 +33,56 @@ You can read more about the Infrastructure Data Standard in [CoST Guidance Note
-```eval_rst
-.. admonition:: Mapping from OCDS
-    :class: Note
+The CoST IDS also sets out a number of disclosure requirements under the heading of 'information for disclosure upon request', also known as 'reactive disclosure'. You can disclose these elements proactively using OC4IDS.
````
Should we link to the reactive disclosure tables here, or otherwise clarify that proactive and reactive mappings are set out separately (as they are in the IDS)?
 
And maybe update the **Disclosure** section on line 12 which only notes 40 items of proactive data?

duncandewhurst replied:
I've added the clarification in https://github.com/open-contracting/infrastructure/pull/295/commits/68ca204e39c4afbbc1b154acc4d4232bb7cbfde2.

I'll leave line 12 as it is, since the reactive items are introduced with 'also'

pindec commented on mapping/reactive-project-level-identification-preparation.csv:
```diff
@@ -0,0 +1,46 @@
+
+* Set its ``.period.endDate`` to the last day of the year, e.g ‘2020-12-31T23:59:59Z",*Copy the contents of* ``/planning/budget/budgetBreakdown`` *in each contracting process to* ``budget.BudgetBreakdown``.
+Project brief or Feasibility study,"Documentation of feasibility studies carried out for this contracting process or project, providing information on net benefits or costs of the proposed goods, works or services.",".. markdown::
+  [Add a project document](#add-a-project-document)and set ``.documentType`` to 'feasibilityStudy'.",*Check the* ``/planning/documents`` *array for* ``Document`` *objects with* ``.type`` *set to* ‘feasibilityStudy’ *and copy each object to the* ``documents`` *array.*
```
Add space after close bracket -document)and

duncandewhurst replied:
Fixed in https://github.com/open-contracting/infrastructure/pull/295/commits/68ca204e39c4afbbc1b154acc4d4232bb7cbfde2

pindec commented on mapping/reactive-project-level-identification-preparation.csv:
```diff
@@ -0,0 +1,46 @@
+
+* Set its ``.name`` to the project official’s name
+
+* Set its ``.jobTitle`` to the project official’s job title","Check the ``/parties`` array for ``Organization`` objects with 'procuringEntity' in ``/roles``. Copy each object to the ``parties`` array in OC4IDS. If there is more than one procuring entity, do not set ``.summary.tender.procuringEntity.name`` and ``.summary.tender.procuringEntity.id``; otherwise, set them to the procuring entity's `` ...
```
Here it looks like we're using slash notation ``/keyword`` to indicate relative navigation of OCDS data, and ``fieldnames`` to indicate fields in OC4IDS?  

I think there's a consistency issue with other sections about which we use - e.g. in  /en/cost/#project-completion Mapping to OC4IDs uses slashes (e.g. `completion/finalValue`) and Mapping from OCDS uses dot notation (`contracts.value`).

Maybe we can be consistent if we also use colour to indicate which we are talking about?

duncandewhurst replied:
Ah yes, with the new additions I was trying to follow https://ocds-standard-development-handbook.readthedocs.io/en/latest/meta/style_guide.html#text-formatting. I've updated the other mappings to be consistent in https://github.com/open-contracting/infrastructure/pull/295/commits/68ca204e39c4afbbc1b154acc4d4232bb7cbfde2

pindec commented on mapping/process-level-procurement.csv:
```diff
@@ -1,14 +1,18 @@
+Number of firms tendering,Number of firms who submit a tender,Contracting process: Publish in ``.summary.tender.numberOfTenderers``,Set ``.summary.numberOfTenderers`` to the value of ``tender.numberOfTenderers``.
+Cost estimate,Currency and amount of the original pre-tender estimate of the contract,Contracting process: Publish in ``.summary.tender.costEstimate``,Check the individual releases for the contracting process and set ``contractingProcess.summary.tender.costEstimate`` to the value of ``tender.value`` in the most recent release with `tender.status`` set to 'planning'.
+Contract title,The formal name of the contract,Contracting process: Publish in ``.summary.title``,"If the contracting process has only one ``Contract`` object in ``contracts``, set ``.summary.title`` to the value of the contract's ``.title``; if not available and the contracting process has only one ``Award`` object in ``awards``, use its ``.title``. Otherwise, use ``tender.title``."
+Contract firm(s),Legal name of supplier,Project Level: Add an entries to ``parties`` with 'supplier' included in their ``.roles`` | Contracting process: Add the names and identifiers to the ``.summary.suppliers`` array,"Check the ``parties`` array for ``Organization`` objects with 'supplier' in ``.roles``. For each such object, copy it to the ``parties`` array in OC4IDS, add an ``OrganizationRefe ...
```
Might be easier to read if we omit the 's on ``OrganizationReference`` and ``Organization``.

pindec replied:
Also, ``.summary.suppliers` -> ``.summary.suppliers``

duncandewhurst replied:
Done in https://github.com/open-contracting/infrastructure/pull/295/commits/68ca204e39c4afbbc1b154acc4d4232bb7cbfde2

pindec commented on mapping/reactive-process-level-implementation.csv:
```diff
@@ -0,0 +1,33 @@
+
+  [Add a contracting process document](#add-a-contracting-process-document) and set ``.documentType`` to 'paymentCertificate'.","*Copy the contents of the* ``/contracts/implementation/transactions`` *array to* ``.summary.transactions``.
+
+*Check the* ``/contracts/implementation/documents`` *array for* ``Document`` *objects with* ``.type`` *set to* ‘qualityAssuranceReport’ *and copy each object to the* ``.summary.documents`` *array.*"
```
Should this documentType ‘qualityAssuranceReport’ code be 'paymentCertificate'?

duncandewhurst replied:
Yes :-) fixed in https://github.com/open-contracting/infrastructure/pull/295/commits/68ca204e39c4afbbc1b154acc4d4232bb7cbfde2

pindec commented on mapping/process-level-procurement.csv:
```diff
@@ -1,14 +1,18 @@
+Contract title,The formal name of the contract,Contracting process: Publish in ``.summary.title``,"If the contracting process has only one ``Contract`` object in ``contracts``, set ``.summary.title`` to the value of the contract's ``.title``; if not available and the contracting process has only one ``Award`` object in ``awards``, use its ``.title``. Otherwise, use ``tender.title``."
+Contract firm(s),Legal name of supplier,Project Level: Add an entries to ``parties`` with 'supplier' included in their ``.roles`` | Contracting process: Add the names and identifiers to the ``.summary.suppliers`` array,"Check the ``parties`` array for ``Organization`` objects with 'supplier' in ``.roles``. For each such object, copy it to the ``parties`` array in OC4IDS, add an ``OrganizationRefe ...
+Contract price,Currency and price at contract award,Contracting process: Publish in ``.summary.contractValue``,"If the value of ``value.currency`` is the same for all ``Award`` objects in ``awards``, set ``.summary.contractValue.amount`` to the sum of ``value.amount`` in each ``Award`` object and set ``.summary.contractValue.currency`` to the value of ``value.currency``.  *Otherwise, convert* ``v ...
+Contract scope of work,"Main outputs from the contract e.g. detailed design, supervision, project management and or type, quantity, unit for construction",Contracting process: Publish in ``.summary.description``. Additionally detailed documentation may be provided as documents within linked releases,"If there is more than one ``Contract`` object in ``contracts`` or more than one ``Award`` object  ...
```
to the value of ``tender.description``` -> to the value of ``tender.description``

duncandewhurst replied:
Fixed in https://github.com/open-contracting/infrastructure/pull/295/commits/68ca204e39c4afbbc1b154acc4d4232bb7cbfde2

pindec reviewed (approved):
Thanks @duncandewhurst. A couple of minor typos, plus two minor consistency points (I didn't flag all the lines):
1. Using slash or dot notation consistently to indicate OCDS and/or OC4IDS fields
2. Having a full stop (or not) at the end of each sentence in the mapping tables.

duncandewhurst committed:
Update mapping tables and documentation

Use consistent punctuation. Fix formatting errors. Add links. Use consistent notation to refer to fields in OCDS data and fields in OC4IDS schema.

duncandewhurst commented:
Thanks, @pindec!

@jpmckinney this is ready for your review now

jpmckinney reviewed (approved):
I haven't reviewed the CSVs or JSON examples, as I assume they received detailed attention.

I changed the base to 210 to make it easier to review. It can be merged into that PR, or the base can be changed back.

duncandewhurst committed:
Merge branch '0.9-dev' into 220-reactive-mappings
